File: bmlib/db/transactions.py
# bmlib — shared library for biomedical literature tools
# Copyright (C) 2024-2026 Dr Horst Herb
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""Transaction context manager."""

from __future__ import annotations

import logging
import threading
from collections.abc import Generator
from contextlib import contextmanager
from typing import Any

from bmlib.db.backend import is_sqlite

logger = logging.getLogger(__name__)

_SAVEPOINT = "bmlib_transaction"

# How many :func:`transaction` blocks are currently open on a connection,
# keyed by ``id(conn)``. Only PostgreSQL needs this: psycopg2 opens a
# transaction on the *first statement of any kind* — a bare ``SELECT`` leaves
# the connection INTRANS — so the driver's own transaction status cannot tell
# "someone called transaction() around me" from "someone ran a query". Getting
# that wrong would silently stop committing. SQLite is not tracked here; its
# ``conn.in_transaction`` answers the same question directly.
#
# The counter is keyed by id() rather than stored on the connection because
# psycopg2's connection is a C type that rejects attribute assignment. Entries
# are removed as the outermost block exits, so an id is never left registered
# for a connection that has been garbage collected.
_depths: dict[int, int] = {}
_depths_lock = threading.Lock()


def transaction_depth(conn: Any) -> int:
    """Return how many :func:`transaction` blocks are open on *conn*.

    Zero means the next :func:`transaction` block owns the commit.
    """
    with _depths_lock:
        return _depths.get(id(conn), 0)


@contextmanager
def _depth_tracked(conn: Any) -> Generator[None, None, None]:
    """Count one open :func:`transaction` block for *conn*."""
    key = id(conn)
    with _depths_lock:
        _depths[key] = _depths.get(key, 0) + 1
    try:
        yield
    finally:
        with _depths_lock:
            remaining = _depths.get(key, 1) - 1
            if remaining > 0:
                _depths[key] = remaining
            else:
                _depths.pop(key, None)


def owns_commit(conn: Any) -> bool:
    """Return True if a write on *conn* right now would need its own commit.

    False means the caller is inside a :func:`transaction` block that will
    commit on its way out, so an inner helper must not commit on its own.
    """
    return transaction_depth(conn) == 0


def _is_nested(conn: Any) -> bool:
    """Return True if this :func:`transaction` block is inside another one."""
    if is_sqlite(conn):
        # sqlite3 auto-begins only before DML, so pending writes here really do
        # mean an enclosing transaction whose owner will commit.
        return bool(conn.in_transaction)
    return transaction_depth(conn) > 0


def _run(conn: Any, sql: str) -> None:
    """Execute a bare statement on either backend.

    ``sqlite3.Connection`` has a convenience ``execute()``; psycopg2's does
    not, so savepoint control has to go through a cursor there.
    """
    if is_sqlite(conn):
        conn.execute(sql)
    else:
        conn.cursor().execute(sql)


@contextmanager
def transaction(conn: Any) -> Generator[Any, None, None]:
    """Context manager that commits on success, rolls back on exception.

    Usage::

        with transaction(conn):
            execute(conn, "INSERT INTO ...")
            execute(conn, "UPDATE ...")
        # auto-committed here

    For SQLite, ``conn.execute("BEGIN")`` is issued explicitly so that
    ``conn.commit()`` has a well-defined scope.  For PostgreSQL (psycopg2),
    autocommit is off and a transaction begins implicitly with the first
    statement, so the outermost block just commits or rolls back.

    Nesting: entering a block while another is already open runs the inner one
    inside a ``SAVEPOINT``. On exception only the inner block's writes are
    rolled back — the outer block's pending writes survive, still uncommitted.
    On success the savepoint is released and **no commit is issued**: whoever
    opened the outermost block owns the commit. This is what makes nesting
    composable — a batch loop can wrap many ``transaction()``-using calls in
    one outer ``transaction()`` and pay a single commit, and an outer failure
    rolls back the inner blocks' writes too. :func:`bmlib.publications.sync`
    depends on it for one-commit-per-day batching.

    How "already open" is decided differs by backend, and deliberately so.
    SQLite auto-begins only before DML, so ``conn.in_transaction`` means what
    it says. psycopg2 begins a transaction on the first statement of any kind
    — a bare ``SELECT`` is enough — so its transaction status would report
    "already open" for a connection nobody has wrapped, and every write would
    quietly stop committing. PostgreSQL therefore counts bmlib's own open
    blocks (see :func:`transaction_depth`) instead of asking the driver.

    Reusing one savepoint name at every level is safe: ``ROLLBACK TO`` and
    ``RELEASE`` address the *most recent* savepoint of that name, which —
    because the blocks are strictly nested — is always this block's own.
    """
    if _is_nested(conn):
        # Join the enclosing transaction via a savepoint so an exception rolls
        # back only this block's writes (see docstring).
        _run(conn, f"SAVEPOINT {_SAVEPOINT}")
        with _depth_tracked(conn):
            try:
                yield conn
            except Exception:
                _run(conn, f"ROLLBACK TO SAVEPOINT {_SAVEPOINT}")
                _run(conn, f"RELEASE SAVEPOINT {_SAVEPOINT}")
                raise
            _run(conn, f"RELEASE SAVEPOINT {_SAVEPOINT}")
        return

    if is_sqlite(conn):
        conn.execute("BEGIN")

    with _depth_tracked(conn):
        try:
            yield conn
            conn.commit()
        except Exception:
            conn.rollback()
            raise

File: spikes/db-rs/src/transactions.rs
// bmlib — shared library for biomedical literature tools
// Copyright (C) 2024-2026 Dr Horst Herb
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

//! Composable transactions via savepoints.
//!
//! The Python contract this reproduces, verbatim from `bmlib/db/transactions.py`:
//! entering a block while another is open runs the inner one inside a
//! `SAVEPOINT`; on success the savepoint is released and **no commit is
//! issued**, because whoever opened the outermost block owns the commit; on
//! failure only the inner block's writes are rolled back.
//!
//! What is *not* reproduced is the machinery that decided "am I nested?" —
//! `_depths`, `_depth_key`, `_depths_lock`, `_is_nested`, `transaction_depth`
//! and the `(thread, id(conn))` keying, about 60 lines of the Python module
//! and the subject of its longest comment. Here the answer is the type of the
//! thing you were handed, so there is nothing to look up and no lock to take.

use crate::db::Db;
use crate::error::Result;

/// Run `f` in a transaction, committing on `Ok` and rolling back on `Err`.
///
/// Nested calls open a savepoint instead, exactly as the Python original does.
///
/// # Why a closure and not a guard
///
/// `Drop` cannot return a `Result` and cannot see whether the block succeeded,
/// so an RAII guard can only auto-*rollback*; auto-commit-on-success needs the
/// success/failure signal a closure's return value carries. This is the one
/// place the Rust API reads differently from `with transaction(conn):`, and it
/// is strictly stronger: a failing commit is reported rather than swallowed.
///
/// # Example
///
/// A helper written against `&mut dyn Db` composes in either position — the
/// property `publications.sync()`'s one-commit-per-day batching depends on:
///
/// ```
/// use bmlib_db::{open_memory, transaction, params, operations::{create_tables, execute}};
///
/// fn store(db: &mut dyn bmlib_db::Db, v: &str) -> bmlib_db::Result<()> {
///     transaction(db, |tx| {
///         execute(tx, "INSERT INTO t (v) VALUES (?)", &params![v])?;
///         Ok(())
///     })
/// }
///
/// let mut conn = open_memory().unwrap();
/// create_tables(&mut conn, "CREATE TABLE t (v TEXT);").unwrap();
///
/// store(&mut conn, "standalone").unwrap();              // a real transaction
/// transaction(&mut conn, |tx| {
///     store(tx, "batched")?;                            // a savepoint
///     store(tx, "batched too")?;                        // another savepoint
///     Ok(())
/// }).unwrap();                                          // one commit for both
/// ```
///
/// # The hazard the borrow checker removes
///
/// Python's depth table exists because the *same* `conn` object is passed into
/// helpers whether or not a block is open, so only a side count can tell an
/// outermost block from a nested one — and getting the key wrong (by
/// connection alone rather than by thread *and* connection) silently stopped
/// committing. Here, reaching around an open block to the connection it
/// borrows does not compile:
///
/// ```compile_fail
/// use bmlib_db::{open_memory, transaction, params, operations::execute};
///
/// let mut conn = open_memory().unwrap();
/// transaction(&mut conn, |_tx| {
///     // error: cannot borrow `conn` as mutable more than once at a time
///     execute(&mut conn, "INSERT INTO t (v) VALUES (?)", &params!["x"])?;
///     Ok(())
/// }).unwrap();
/// ```
pub fn transaction<T, F>(db: &mut dyn Db, f: F) -> Result<T>
where
    F: FnOnce(&mut dyn Db) -> Result<T>,
{
    let mut block = db.begin()?;
    match f(&mut *block) {
        Ok(value) => {
            block.commit()?;
            Ok(value)
        }
        Err(err) => {
            // The caller's error is what matters; a rollback that also fails
            // must not mask it. Python has the same precedence — `raise` in
            // the `except` clause runs after `conn.rollback()`.
            let _ = block.rollback();
            Err(err)
        }
    }
}

/// True if a write on `db` right now would need its own commit.
///
/// Kept because it is public API in the Python original. Its only consumer
/// there — `create_tables` — no longer needs it: a `Connection` is autocommit
/// and a `Transaction` is owned by its opener, so neither has a decision to
/// make. See FINDINGS.md, "What disappears".
pub fn owns_commit(db: &dyn Db) -> bool {
    db.owns_commit()
}
